### PR TITLE
Feature/wrapper

### PR DESCRIPTION
**Pull request by ddelpiano**
Wrapper implementation:

- wrapper/bin/converter: quite minimal script that instantiate the conversion handler and get id/file and method.
- wrapper/bin/model/filehandler: generic class that iterate through all the format supported and go step by step in the conversion, plus there is the id retrieval from google sheet.
- All the logic that required to interact with the pre-existent model has been implemented in modelClasses, where I have kind of deserialiser for each step of the conversion required.

Known issue: right now I can still see few discrepancy from generated to JsonLD, still trying to assess them.

### CODE REVIEW

File: wrapper/bin/model/filehandler.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            return _filename;
+        });
+    }
+
+    #fromXlsxToJson(file) {
+        try {
+            if (fs.existsSync(file)) {
+                let filename = this._destination_folder + "/model.json";
+                let _xlsx = fs.readFileSync(file, 'binary');
+                let _json_model = converter.fromXLSXToJson(_xlsx);
```
Comment (zsinnema) [inline wrapper/bin/model/filehandler.js:73]: const?

File: wrapper/bin/model/filehandler.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            fs.writeFileSync(_filename, res.data);
+            return _filename;
+        });
+    }
+
+    #fromXlsxToJson(file) {
+        try {
+            if (fs.existsSync(file)) {
+                let filename = this._destination_folder + "/model.json";
+                let _xlsx = fs.readFileSync(file, 'binary');
```
Comment (zsinnema) [inline wrapper/bin/model/filehandler.js:72]: const?

File: wrapper/bin/model/filehandler.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        }
+    }
+
+    #fromLDToFlattened(file) {
+        try {
+            if (fs.existsSync(file)) {
+                var filename = this._destination_folder + "/model-flattened.jsonLD";
+                var _jsonld = fs.readFileSync(file, 'binary');
+                const _callback = function callback(res) {
+                    let _flattened = JSON.stringify(res, null, 2);
```
Comment (zsinnema) [inline wrapper/bin/model/filehandler.js:122]: const

File: wrapper/bin/model/filehandler.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            } else {
+                throw new Error('The file given in input does not exist or is not located where specified.')
+            }
+        } catch(err) {
+            throw new Error('An error has been encoutered during the conversion from XLSX to Json.')
+        }
+    }
+
+    async convertAll() {
+        var startConverting = false;
```
Comment (zsinnema) [inline wrapper/bin/model/filehandler.js:136]: let

File: wrapper/package.json [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  "description": "wrapper for apinatomy angular application",
+  "main": "bin/converter.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "Dario Metacell",
+  "bin": {
+    "wrapper": "./bin/converter.js"
+  },
+  "license": "ISC",
```
Comment (tgbugs) [inline wrapper/package.json:13]: Can this be Apache2? re: https://github.com/open-physiology/open-physiology-viewer/issues/66

**Comment by tgbugs** [comment]
Two requests and two issues I have encountered while testing.

Requests.
1. Rename `-m --method` to `-f --from` and add an option `-t --to` which will output only the single stage of conversion. If `-t` is not specified then a folder with all the subsequent formats is produced as it is now.
2. Can we get a `-o --output` option that specifies the name of the output file or folder?

For the issues, both were resolved by commenting out the offending variables. I'm not sure why those were being pulled in to the top level.

```
/home/tom/git/open-physiology-viewer/dist/converter.js:5
/******/ 	var parentHotUpdateCallback = window["webpackHotUpdate"];
         	                              ^

ReferenceError: window is not defined
    at /home/tom/git/open-physiology-viewer/dist/converter.js:5:41
    at Object.<anonymous> (/home/tom/git/open-physiology-viewer/dist/converter.js:860:10)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (/home/tom/git/open-physiology-viewer/wrapper/bin/model/filehandler.js:3:19)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
```

```
/home/tom/git/open-physiology-viewer/dist/converter.js:61308
const crypto = self.crypto || self.msCrypto;
               ^

ReferenceError: self is not defined
    at Object.../node_modules/rdf-canonize/lib/MessageDigest-browser.js (/home/tom/git/open-physiology-viewer/dist/converter.js:61308:16)
    at __webpack_require__ (/home/tom/git/open-physiology-viewer/dist/converter.js:783:30)
    at fn (/home/tom/git/open-physiology-viewer/dist/converter.js:94:20)
    at Object.<anonymous> (/home/tom/git/open-physiology-viewer/dist/converter.js:61870:23)
    at Object.../node_modules/rdf-canonize/lib/URDNA2015.js (/home/tom/git/open-physiology-viewer/dist/converter.js:62375:30)
    at __webpack_require__ (/home/tom/git/open-physiology-viewer/dist/converter.js:783:30)
    at fn (/home/tom/git/open-physiology-viewer/dist/converter.js:94:20)
    at Object.../node_modules/rdf-canonize/lib/index.js (/home/tom/git/open-physiology-viewer/dist/converter.js:63121:19)
    at __webpack_require__ (/home/tom/git/open-physiology-viewer/dist/converter.js:783:30)
    at fn (/home/tom/git/open-physiology-viewer/dist/converter.js:94:20)

```

**Comment by ddelpiano** [comment]
> Two requests and two issues I have encountered while testing.
> 
> Requests.
> 
> 1. Rename `-m --method` to `-f --from` and add an option `-t --to` which will output only the single stage of conversion. If `-t` is not specified then a folder with all the subsequent formats is produced as it is now.
> 2. Can we get a `-o --output` option that specifies the name of the output file or folder?
> 
> For the issues, both were resolved by commenting out the offending variables. I'm not sure why those were being pulled in to the top level.
> 
> ```
> /home/tom/git/open-physiology-viewer/dist/converter.js:5
> /******/ 	var parentHotUpdateCallback = window["webpackHotUpdate"];
>          	                              ^
> 
> ReferenceError: window is not defined
>     at /home/tom/git/open-physiology-viewer/dist/converter.js:5:41
>     at Object.<anonymous> (/home/tom/git/open-physiology-viewer/dist/converter.js:860:10)
>     at Module._compile (internal/modules/cjs/loader.js:1063:30)
>     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
>     at Module.load (internal/modules/cjs/loader.js:928:32)
>     at Function.Module._load (internal/modules/cjs/loader.js:769:14)
>     at Module.require (internal/modules/cjs/loader.js:952:19)
>     at require (internal/modules/cjs/helpers.js:88:18)
>     at Object.<anonymous> (/home/tom/git/open-physiology-viewer/wrapper/bin/model/filehandler.js:3:19)
>     at Module._compile (internal/modules/cjs/loader.js:1063:30)
> ```
> 
> ```
> /home/tom/git/open-physiology-viewer/dist/converter.js:61308
> const crypto = self.crypto || self.msCrypto;
>                ^
> 
> ReferenceError: self is not defined
>     at Object.../node_modules/rdf-canonize/lib/MessageDigest-browser.js (/home/tom/git/open-physiology-viewer/dist/converter.js:61308:16)
>     at __webpack_require__ (/home/tom/git/open-physiology-viewer/dist/converter.js:783:30)
>     at fn (/home/tom/git/open-physiology-viewer/dist/converter.js:94:20)
>     at Object.<anonymous> (/home/tom/git/open-physiology-viewer/dist/converter.js:61870:23)
>     at Object.../node_modules/rdf-canonize/lib/URDNA2015.js (/home/tom/git/open-physiology-viewer/dist/converter.js:62375:30)
>     at __webpack_require__ (/home/tom/git/open-physiology-viewer/dist/converter.js:783:30)
>     at fn (/home/tom/git/open-physiology-viewer/dist/converter.js:94:20)
>     at Object.../node_modules/rdf-canonize/lib/index.js (/home/tom/git/open-physiology-viewer/dist/converter.js:63121:19)
>     at __webpack_require__ (/home/tom/git/open-physiology-viewer/dist/converter.js:783:30)
>     at fn (/home/tom/git/open-physiology-viewer/dist/converter.js:94:20)
> ```

Hi @tgbugs , sure I will implement 1. and 2.

The issue you had with self and window has been solved already, when I did the cleanup I removed some global declarations that I should have kept in the wrapper, thx for spotting it.

File: src/model/modelClasses.js [hunk 6]
```diff
@@ -119,6 +132,254 @@ export function fromJSON(inputModel) {
...
+
+        model.syncRelationships(modelClasses, entitiesList, namespace);
+        model.entitiesByID = entitiesList;
+    }
+
+    function processScaffoldWaitingList(model, entitiesList) {
+        //Auto-create missing definitions for used references
+        let added = [];
+        (entitiesList.waitingList)::entries().map(([id, refs]) => {
+            let [obj, key] = refs[0];
```
Comment (zsinnema) [inline src/model/modelClasses.js:319]: please (everywhere) use const instead of let for non changing values

File: src/model/modelClasses.js [hunk 7]
```diff
@@ -119,6 +132,254 @@ export function fromJSON(inputModel) {
...
+                        [$Field.generated]: true
+                    })
+
+                    //Include newly created entity to the main graph
+                    let prop = modelClasses[this.name].Model.selectedRelNames(clsName)[0];
+                    if (prop) {
+                        model[prop] = model[prop] || [];
+                        model[prop].push(e);
+                    }
+                    let fullID = getFullID(namespace, e.id);
```
Comment (zsinnema) [inline src/model/modelClasses.js:335]: there is no need to create the fullId variable

File: src/model/modelClasses.js [hunk 8]
```diff
@@ -119,6 +132,254 @@ export function fromJSON(inputModel) {
...
+                    entitiesList[fullID] = e;
+                    added.push(e.id);
+                }
+            }
+        });
+
+        if (added.length > 0) {
+            added.map(id => delete entitiesList.waitingList[id]);
+            let resources = added.filter(id => entitiesList[getFullID(namespace,id)].class !== $SchemaClass.External);
```
Comment (zsinnema) [inline src/model/modelClasses.js:344]: what is the use of this resources?

File: src/model/modelClasses.js [hunk 9]
```diff
@@ -119,6 +132,254 @@ export function fromJSON(inputModel) {
...
+                            }
+                        });
+                    }
+                }
+            }
+        });
+
+        if (added.length > 0){
+            added.map(id => delete entitiesList.waitingList[id]);
+            const resources = added.filter(id => entitiesList[getFullID(namespace,id)].class !== $SchemaClass.External);
```
Comment (zsinnema) [inline src/model/modelClasses.js:307]: what is the use of the resources and externals?

**Review by zsinnema — CHANGES_REQUESTED** [review]

File: src/model/modelClasses.js (hunk 7 above)
Comment (ddelpiano) [reply src/model/modelClasses.js:335]: fullID is used the line below and key for entitiesById map

**Review by zsinnema — APPROVED** [review]
lgtm

**Pull request merged**